trial/src/main/kotlin/com/hartwig/actin/trial/TrialIngestion.kt

```diff
@@ -65,11 +75,30 @@ class TrialIngestion(private val eligibilityFactory: EligibilityFactory) {
                     ),
                     generalEligibility = criteria,
                     cohorts = mappedCohorts
+                )
+                TrialWrapper(
+                    trial = trial,
+                    rules = criteria.map { it.function.rule } +
+                            mappedCohorts.flatMap { cohort -> cohort.eligibility.map { it.function.rule } }
                 ).right()
-            else UnmappableTrial(trialId = trialState.trialId, trialErrors, unmappableCohorts).left()
+            } else {
+                UnmappableTrial(trialId = trialState.trialId, trialErrors, unmappableCohorts).left()
+            }
+        }
+        val (errors, trialWrappers) = trialsAndUnmappableTrials.partitionAndJoin()
+        if (errors.isNotEmpty()) {
+            return errors.left()
+        }
+
+        val (trials, usedRules) = with(trialWrappers) { map { it.trial } to flatMap { it.rules }.distinct() }
+        val eligibilityRulesState = with(EligibilityRule.entries.associateWith { EligibilityRuleUsedStatus.UNUSED }.toMutableMap()) {
+            usedRules.forEach { this[it] = EligibilityRuleUsedStatus.USED }
+            this.map { EligibilityRuleState(it.key, it.value) }
         }
```

Comment:
Could this be simplified to something like:

```suggestion
EligibilityRule.entries.map { rule ->
            EligibilityRuleState(rule, if (rule in usedRules) EligibilityRuleUsedStatus.USED else EligibilityRuleUsedStatus.UNUSED)
        }
 ```

Reply:
Yes, but it's more complicated (in terms of O complexity). But I think readability is more important than optimization here:) So, fixed!